Public code release

GabPrato opened the pull request.

GabPrato committed:
Copied code from private to public repo.

AndreasMadsen commented on accelerate_patch/apply_patch.sh:
```diff
@@ -0,0 +1,10 @@
+# This script is meant to be ran from the root of the repo: `bash accelerate_patch/apply_patch.sh`
+
+# Locate the accelerate library directory
+ACCELERATE_DIR=$(python -c "import accelerate; import os; print(os.path.dirname(accelerate.__file__))")
```
If using a patch is really the intended solution. Then you should pin accelerate to a specific commit.

GabPrato replied:
The "patch" is not meant to fix an issue with accelerate's DataLoader, rather it adds support for the type of batch expected by our benchmark.

AndreasMadsen commented on pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+]
+
+[project.urls]
+Homepage = "http://chandar-lab.github.io/EpiK-Eval"
```
Not found.

GabPrato replied:
Fixed

AndreasMadsen commented on pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+requires = ["setuptools"]
+build-backend = "setuptools.build_meta"
+
+[project]
```
I don't think this setup is correct. You appear to use the `src-layout` (https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#src-layout) but have not places your content in a named subdirectory. I would guess this code installs 3 root packages named `__init__`, `dataloader`, and `run`. When it properly should install just one package named `epik_eval`.

Also, consider using a `__main__.py` instead of a `run.py`.

GabPrato replied:
Fixed

AndreasMadsen reviewed (commented):
**This is not a complete review. I can review again after the refactor.**

I don't think `dataset_generator`, `generate_paper_plots`, and `compute_paper_metrics`, are well suited as notebooks. They appear to be a collection of several independent scripts. I think the code would be better structured as such.

It's also not possible for me to properly review that code, with the GitHub interface as it's notebooks.

The README is missing instructions regarding the notebook code. Which appears to be essential to generate the paper results.

Regarding the accelerator patch. Have you tried upstreaming the changes? If not, you should try. If yes, maybe it's better to point to your repo-brach as a dependency. It would be less confusing.

GabPrato committed:
Converted dataset generator notebook to python script.

GabPrato committed:
Put dataset and names.txt in data/

GabPrato committed:
Moved code in src/ to src/epik_eval/

GabPrato committed:
Updated README

GabPrato committed:
Added setup.py to create logs/ dir during setup

GabPrato committed:
Replaced notebooks/ with scripts/

GabPrato committed:
Fixed new data/ dir in Makefile

GabPrato committed:
Added pandas to requirements

GabPrato committed:
Fixed data/ path in dataloader.py

GabPrato committed:
Added optional matplotlib to requirements.

GabPrato committed:
Fixed dataset generation script minor bugs.

GabPrato committed:
Fixed typo in paper plot generation script

GabPrato committed:
Specified that scripts are meant to be ran in the scripts/ dir

GabPrato committed:
Minor fixes to compute_paper_metrics.py script

GabPrato commented:
@AndreasMadsen addressed all mentioned issues. Code is ready for a second review.

GabPrato committed:
Fixed text centering in README.md

GabPrato committed:
Update README.md

AndreasMadsen commented on setup.py:
```diff
@@ -0,0 +1,20 @@
+import os
```
You should not have both a `setup.py` and `pyproject.toml` file. `pyproject.toml` have replaced `setup.py` in modern python.

AndreasMadsen commented on setup.py:
```diff
@@ -0,0 +1,20 @@
+class AddLogsDir(install):
+    def run(self):
+        # Add logs/ directory for experiments
+        os.makedirs('logs', exist_ok=True)
```
You can make directories before the file is written, not on install. Install side-effects are deprecated due to the security risk.

AndreasMadsen commented on accelerate_patch/accelerator.patch:
```diff
@@ -0,0 +1,25 @@
+ 
+-            batch_size_per_device = min(batch_sizes) if deepspeed_plugin.is_train_batch_min else max(batch_sizes)
++            #batch_size_per_device = min(batch_sizes) if deepspeed_plugin.is_train_batch_min else max(batch_sizes)
++            batch_size_per_device = batch_sizes[0] # The first is always the train batch size, the second is the eval batch size.
```
Why is this required? Should the training batch_size not be less than the eval batch size?

GabPrato replied:
Expecting the train batch size to always be smaller than the eval batch size isn't reliable. Although you could argue that expecting the train batch size to always be first in the list of batch sizes is also not reliable. Given that it is in my code though, this is more robust.

AndreasMadsen commented on accelerate_patch/accelerator.patch:
```diff
@@ -0,0 +1,25 @@
+         )
+ 
+         config_kwargs = {
+-            "train_micro_batch_size_per_gpu": batch_size_per_device,
```
I think you can just set `batch_sizes[0] = batch_sizes[0] // deepspeed_plugin.deepspeed_config["gradient_accumulation_steps"]` when configuring accelerate. The values will be the same. This way you can avoid the patch.

GabPrato replied:
I prefer patching for the reason mentioned above. While at it, might as well patch this part.

AndreasMadsen commented on accelerate_patch/data_loader.patch:
```diff
@@ -0,0 +1,55 @@
++            # Take batch slice corresponding to process
++            assert (batch[0].shape[0] % self.state.num_processes) == 0
++            device_batch_size = batch[0].shape[0] // self.state.num_processes
++            device_batch = [tensor[self.state.process_index * device_batch_size : (self.state.process_index + 1) * device_batch_size] for tensor in batch]
```
All this code is just transforming batch. It's unclear why it cannot be implemented with `slice_fn_for_dispatch `. If not, you should open an issue or contribute a PR.

GabPrato replied:
I need to patch `data_loader.py` for the code above this. While at it, having this code here or in `__main__.py` makes no difference to me.

AndreasMadsen commented on configs/flan-t5-base/run_segmented.sh:
```diff
@@ -0,0 +1,17 @@
+accelerate launch --config_file=./configs/flan-t5-base/accelerate_config.yaml src/epik_eval/__main__.py \
```
I don't know accelerate well. Why are some parameters in `.yaml` and others as command line arguments. Should they not all be in `.yaml`?

GabPrato replied:
The `.yaml` contains the config for accelerate, while the command line arguments are for my `__main__.py` script.

AndreasMadsen commented on scripts/generate_dataset.py:
```diff
@@ -0,0 +1,1320 @@
+        if generate:
+            # Val and Test samples
+            for _set in ['val', 'test']:
+                for _ in range(eval(f"args.{_set}_samples_per_task")):
```
use `getattr(args, f'{_set}_samples_per_task')`.

AndreasMadsen commented on scripts/generate_dataset.py:
```diff
@@ -0,0 +1,1320 @@
+            for _set in ['val', 'test']:
+                for _ in range(eval(f"args.{_set}_samples_per_task")):
+                    document_id = len(unsegmented_documents)
+                    unsegmented_document, segmented_document, question, answer = eval(f"task{task}_sample")(character_names)
```
Make an array and index it. `fns = [task1_sample, task2_sample, ...]` and `fns[task - 1]`.

GabPrato replied:
Called `eval` once at the beginning of the main loop instead.

AndreasMadsen commented on scripts/generate_dataset.py:
```diff
@@ -0,0 +1,1320 @@
+                
+            # Q/A examples
+            for _ in range(args.qa_samples_per_task):
+                _, _, question, answer = eval(f"task{task}_sample")(character_names)
```
Same

AndreasMadsen commented on scripts/generate_dataset.py:
```diff
@@ -0,0 +1,1320 @@
+    unsegmented_documents_dir = os.path.join(args.dataset_dir, 'unsegmented_documents/')
+    os.makedirs(unsegmented_documents_dir)
+    for i, unsegmented_doc in enumerate(unsegmented_documents):
+        with open(os.path.join(unsegmented_documents_dir, f'{i:04d}.txt'), 'w') as f:
```
What do these `i` refer to? It looks like they are meant to refer to the task. However, that will only be the case if all tasks are generated. If this is the case, you should fix it such that the indices don't change depending on what tasks are enabled.

GabPrato replied:
They are the doc's id. I renamed the variable.

AndreasMadsen commented on scripts/generate_paper_plots.py:
```diff
@@ -0,0 +1,1445 @@
+                            test_t5_unsegmented, test_t5_segmented, test_flan_t5_unsegmented, test_flan_t5_segmented, test_opt_unsegmented, test_opt_segmented,
+                            t5_sizes, opt_sizes,
+                            filename):
+    plt.close()
```
Side-effect. Better to close the plot in the caller or the function that created the plot that is now being closed.

AndreasMadsen commented on scripts/generate_paper_plots.py:
```diff
@@ -0,0 +1,1445 @@
+    opt_unsegmented = [46.89, 53.44, 57.78, 65.44]
+    opt_segmented = [7.00, 7.33, 12.22, 12.89]
+
+    plot_accuracy(t5_unsegmented, t5_segmented, flan_t5_unsegmented, flan_t5_segmented, opt_unsegmented, opt_segmented, 
```
Not a required change since they are just plot. But it's usually nice to be able to generate just one plot. As generating all is hard to debug. I suggest code for each plot is moved to it's own file.

AndreasMadsen commented on scripts/README.md:
```diff
@@ -0,0 +1,4 @@
+The following scripts can be used to reproduce the results from the paper. Each script is meant to be ran in the `/scripts/` directory, e.g., `python generate_paper_plots.py`.
+* **compute_paper_metrics.py:** Processes a log of model answers generated by running the benchmark. It then outputs the different metrics detailed in the paper.
+* **generate_dataset.py:** Used to generate a dataset for our benchmark. Optional, since we already provide one in `/data/`.
```
The dataset is core to the repo and the code. You should move the docs to the main README and document it more.

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+        train_tokenizer = AutoTokenizer.from_pretrained(args.model_name)
+        eval_tokenizer = AutoTokenizer.from_pretrained(args.model_name, padding_side='left') # For correct generation, padding must be left
+        if train_tokenizer.pad_token_id is None:
+            train_tokenizer.add_special_tokens({'pad_token': '[PAD]'})
```
If `pad_token` are not already part of the model. Why would the token mask be handled correctly?

GabPrato replied:
The tokenizer does mask these tokens.

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+            batch, attention_mask, seq_length, target, target_length, _ = batch
+
+        # Iterate minibatches
+        for i in range(0, batch.shape[0], args.minibatch_size):
```
Why does the `train_dataloader` not produce mini-batches? If there is a good reason, you should document it.

GabPrato replied:
The minibatch size is per device. If I were to initialize the dataloader with the minibatch size instead of the total batch size, it would divide the minibatch size by the number of device, which would change how I defined it and thus its value in each config.

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+    # Evaluation
+    results = []
+    while True:  # We have to iterate like this, because if not, the Accelerate dataloader merges samples from different batches, which are of different lengths and thus crashes.
+        prompt_id, prompt, attention_mask, nonduplicate_mask = next(iter(eval_dataloader))
```
If I understand correctly. This will give the same response every time. Calling 'iter()' is supposed to be a stateless operation. If it is stateless, why not cache it outside the while loop? If it is not stateless, this is an anti-pattern that should be fixed.

GabPrato replied:
I've explained the design decisions all the way below this pull request. Let me know if you still have comments or questions.

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+        results.append((prompt_id, output, nonduplicate_mask))
+
+        # Check if main process is finished iterating questions
+        if accelerator.gather(torch.tensor(eval_dataloader.dataset.finished, device=device)).any():
```
Why `any()` should it not be `all()` or check only the main process?
Could you not check if `next(iter)` throws `StopIteration`?

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+        dim=1,
+        pad_index=eval_tokenizer.pad_token_id,
+        pad_first=eval_tokenizer.padding_side == 'left')
+    results = accelerator.gather(results)
```
I think you only want to gather in the main process, right?

GabPrato replied:
`gather` needs to be called on all devices, or else it blocks.

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+
+        accuracies = {}
+        for set in ['train', 'val', 'test']:
+            questions = model_answers_csv[model_answers_csv['set'] == set]
```
`pd.DataFrame.query` is pretty good for this.

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+            accuracies[f'{set}_per_task'] = acc_per_task
+
+        csv_path = args.model_answers_csv.replace('.csv', f'_epoch{epoch}.csv')
+        pathlib.Path(csv_path).parents[0].mkdir(parents=True, exist_ok=True)
```
Minor: replace `parents[0]` with `parent`.

AndreasMadsen commented on src/epik_eval/__main__.py:
```diff
@@ -0,0 +1,483 @@
+                acc_per_task[f'{set.capitalize()} Task {task} Accuracy'] = 100 * task_questions['correct'].sum() / len(task_questions['correct'])
+            accuracies[f'{set}_per_task'] = acc_per_task
+
+        csv_path = args.model_answers_csv.replace('.csv', f'_epoch{epoch}.csv')
```
Minor: since you are already using path lib, you can use `path.with_stem(f"{path.stem}_epoch{epoch}")`.

AndreasMadsen commented on src/epik_eval/dataloader.py:
```diff
@@ -0,0 +1,205 @@
+        answers = qa_df['answer'].to_list()
+
+        # Model is also evaluated for its ability to memorize training samples.
+        train_doc_ids = []
```
It's suspicious that these are called train when it's an eval dataloader.

GabPrato replied:
During evaluation, I also measure how well the model memorized training samples.

AndreasMadsen commented on src/epik_eval/dataloader.py:
```diff
@@ -0,0 +1,205 @@
+            shuffle=False)
+
+    def __iter__(self):
+        for batch in self.dataset:
```
`return iter(self.dataset)` instead. Avoids keeping a generator scope and object.

GabPrato replied:
This is placeholder code, as the dataloader gets replaced by Accelerate's.

AndreasMadsen commented on src/epik_eval/dataloader.py:
```diff
@@ -0,0 +1,205 @@
+        self.finished = False
+
+    def __iter__(self):
+        if self.finished:
```
An `Iterable` does not finish, it is stateless. An `iterator` finish. I think there is a mix of responsibility here, which is properly what is causing you issues.

I guess, the iterator that this __iter__ returns, should either yield all observations from every batch or it should yield all batches.

GabPrato replied:
See my comment below explaining the design.

AndreasMadsen reviewed (commented):
In addition to documenting the code. I think it would be beneficial to document the dataset. How to decrypt it? What is the format? It likely that someone would not use the code but just the dataset.

You need to document the CLI for creating the dataset in details. Preferably also the metrics and plots (I don't think you need to document that in details).

I terms of the code, I'm very confused about the management of iterables. There seams to be some important dynamic with how batches are managed. I guess it's important that some observations are evaluated together, but it's not clear why. This seams to relate to the accelerator patch, the `while True` and the not-stateless iterable. It's all looks like an anti-pattern. Maybe there is some highly specific reason for all this, but it does not appear to be documented anywhere and I worry it's rather due to a common misunderstand regarding the difference between iterables and iterators.

GabPrato committed:
Added documentation for dataset and scripts.

GabPrato committed:
Fixed missing import

GabPrato committed:
Added *.egg-info to .gitignore

GabPrato committed:
Removed scripts/README.md since now documented in main README.md

GabPrato committed:
Removed setup.py

GabPrato committed:
Minor tweaks and comments for quality and comprehension.

GabPrato committed:
Merge branch 'dev' of https://github.com/chandar-lab/EpiK-Eval into dev

GabPrato commented:
@AndreasMadsen Addressed your latest review. Let me know if you have any further comments or questions.

> In addition to documenting the code. I think it would be beneficial to document the dataset. How to decrypt it? What is the format? It likely that someone would not use the code but just the dataset.

Done.
> You need to document the CLI for creating the dataset in details. Preferably also the metrics and plots (I don't think you need to document that in details).

Done.
> I terms of the code, I'm very confused about the management of iterables. There seams to be some important dynamic with how batches are managed. I guess it's important that some observations are evaluated together, but it's not clear why. This seams to relate to the accelerator patch, the `while True` and the not-stateless iterable. It's all looks like an anti-pattern. Maybe there is some highly specific reason for all this, but it does not appear to be documented anywhere and I worry it's rather due to a common misunderstand regarding the difference between iterables and iterators.

The issue is with the way the accelerate dataloader works. In `EvaluationDataset`, if I yield all of samples at once, Accelerate's dataloader will mix them up when yielding batches to my main script, even if you specify not to shuffle samples. Thus, samples in batches will have different lengths. In my training loop, I solve this by padding all samples up to the same length in `TrainDataset`, and then when I receive a batch in `__main__.py`, I cut up to the longest length, removing the extra padding. For evaluation though, since samples don't need to be shuffled, rather than cutting up to the longest length, it is much more efficient to prepare the batches only once in `EvaluationDataset`, pad to the correct length, and then yield samples only in that batch to Accelerate's Dataloader. This explains the need of a `while True` in the main loop, because now I have to call the __iter__ function once for every batch in my evaluation set.

AndreasMadsen commented:
I'm supported to review the code based on if it's understandable and extendable. Right now, I would say "no" to both.

**Understandable:** The `while True: next(iter(d))` is an anti-pattern. This is not how we iterate on iterables. For example, if `d` was a list, you would just get the first element every time. That is the behavior I expect regardless of which data-structure supports `d`. However, because `d` mutates on `__iter__` we get different responses. You really need to not be mutating in  `__iter__`.

**Extendable:** The code requires patching a specific version of accelerate. That means that someone else can't apply this code to their project and use default accelerate or any other version of accelerate.

---

I would also recommend to never use `eval`, while I can see it properly doesn't present a security risk in this case. It just looks bad, as it's generally an anti-pattern, and half of the our review guidelines are about appearance. It's also slow and makes it harder to debug the code. This is such a minor issue compared to the other two, I don't want to focus on this. But I was surprised you didn't remove them all, so I thought it worth mentioning.

GabPrato committed:
EvaluationDataset now iters whole dataset instead of one batch at a time.

GabPrato committed:
Replaced `eval()` call to `getattr()`

GabPrato commented:
Switched the way the evaluation loop works. Now I pad the entire dataset to the same length, return the length and in the evaluation loop, I cut out the extra padding up to the longest prompt in the current batch.

Also replaced the only remaining `eval()` call.

With respect to the accelerate patch, their `data_loader.py` doesn't do what I want:
```python
if not self._drop_last and stop_iteration and observed_batch_size % self.state.num_processes != 0:
    # If the last batch is not complete, let's add the first batch to it.
    batch = concatenate([batch, first_batch], dim=0)
    # Batch size computation above is wrong, it's off by 1 so we fix it.
    batch_size += 1
```
During evaluation, I don't want to drop the last batch or add the first batch in the last batch. So the only way to avoid this would be to yield additional samples in my evaluation dataset, such that there is enough to "fill" the last batch (`len(eval_dataset) % batch_size == 0`). But then I would need to mark these extra samples as duplicates or samples to be discarded (as I am doing right now, but in the data loader) for the main loop to not take them into consideration. It feels weird to me for a dataset class to handle this, rather than the data loader. I understand that this forces users to stick with a specific accelerate version, but the way I see it, nobody is going to use this code. So I don't think it is worth spending time on this.

AndreasMadsen commented:
According to https://github.com/huggingface/accelerate/issues/125

> For evaluation, you can just truncate the arrays of predictions to the length of your validation dataset, to ignore those added results.

GabPrato committed:
Removed dependency on accelerate patch.

GabPrato commented:
Evaluation dataset now adds duplicates such that the evaluation dataset size is a multiple of the evaluation batch size. Duplicates are eliminated in the main loop as before.

DeepSpeed support has also been removed. All configs weren't using DeepSpeed anyways.

GabPrato committed:
Removed accelerate patch from makefile.

AndreasMadsen reviewed (approved):
I have approved the code. You will need a second review.

For public release this is fine. For anonymous submissions remember to delete the LICENSE and update `pyproject.toml`. 

I think the README could be more clear about how exactly to generate each plot presented in the paper.

jhuang265 reviewed (approved):
LGTM. Like andreas mentioned, maybe replace the figures with the code to generate the figures, but that's not really important.
